packages/protocol/tipmanager/tipmanager.go

```diff
@@ -216,19 +174,171 @@ func (t *TipManager) selectTips(count int) (parents models.BlockIDs) {
 }
 
 // AllTips returns a list of all tips that are stored in the TipManger.
-func (t *TipManager) AllTips() []*scheduler.Block {
-	return t.tips.Keys()
+func (t *TipManager) AllTips() (allTips []*scheduler.Block) {
+	allTips = make([]*scheduler.Block, 0, t.tips.Size())
+	t.tips.ForEach(func(_ models.BlockID, value *scheduler.Block) bool {
+		allTips = append(allTips, value)
+		return true
+	})
+
+	return
 }
 
 // TipCount the amount of tips.
 func (t *TipManager) TipCount() int {
 	return t.tips.Size()
 }
 
+// FutureTipCount returns the amount of future tips per epoch.
+func (t *TipManager) FutureTipCount() (futureTipsPerEpoch map[epoch.Index]int) {
+	futureTipsPerEpoch = make(map[epoch.Index]int)
+	t.futureTips.ForEach(func(index epoch.Index, commitmentStorage *memstorage.Storage[commitment.ID, *memstorage.Storage[models.BlockID, *scheduler.Block]]) {
+		commitmentStorage.ForEach(func(cm commitment.ID, tipStorage *memstorage.Storage[models.BlockID, *scheduler.Block]) bool {
+			futureTipsPerEpoch[index] += tipStorage.Size()
+			return true
+		})
+	})
+
+	return
+}
+
+// PromoteFutureTips promotes to the main tippool all future tips that belong to the given commitment.
+func (t *TipManager) PromoteFutureTips(cm *commitment.Commitment) {
+	t.evictionMutex.RLock()
+	defer func() {
+		t.evictionMutex.RUnlock()
+		t.Evict(cm.Index())
+	}()
+
+	if futureEpochTips := t.futureTips.Get(cm.Index()); futureEpochTips != nil {
+		if tipsForCommitment, exists := futureEpochTips.Get(cm.ID()); exists {
+			tipsToPromote := make(map[models.BlockID]*scheduler.Block)
+			tipsForCommitment.ForEach(func(blockID models.BlockID, tip *scheduler.Block) bool {
+				tipsToPromote[blockID] = tip
+				return true
+			})
+
+			tipsToNotPromote := set.NewAdvancedSet[models.BlockID]()
+			for _, tip := range tipsToPromote {
```

Comment:
Do we actually need this loop? I think we could populate `tipsToNotPromote` in the same loop as `tipsToPromote`?

Reply:
I don't think you actually can. Because in order to know what tips to avoid adding you need to check if in _any_ of the tip has some other tip as a parent; to do so you need first to collect all potential tips, and only then you can iterate through every single one of them and remove their parents from the potential tips set.

Reply:
Yeah, but now we basically iterate twice over the same set of blocks. We could as well collect their parents in the same run and then this third loop stays the same
```
			for tipID, tip := range tipsToPromote {
				if !tipsToNotPromote.Has(tipID) {
					t.addTip(tip)
				}
			}
```